qa/integration-tests/src/test/java/io/camunda/it/migration/util/MigrationITExtension.java

```diff
@@ -0,0 +1,224 @@
+  private Map<String, String> initialEnvOverrides = new HashMap<>();
+  private Map<String, String> upgradeEnvOverrides = new HashMap<>();
+  private final DatabaseType databaseType;
+  private String indexPrefix;
+  private String databaseUrl;
+  private CamundaMigrator migrator;
+  private Path tempDir;
+  private MigrationDatabaseChecks migrationDatabaseChecks;
+  private BiConsumer<DatabaseType, CamundaMigrator> beforeUpgradeConsumer = (db, migrator) -> {};
+
+  public MigrationITExtension() {
+    final String property = System.getProperty(PROP_CAMUNDA_IT_DATABASE_TYPE);
+    databaseType =
+        property == null ? DatabaseType.LOCAL : DatabaseType.valueOf(property.toUpperCase());
+  }
+
+  @Override
+  public void afterAll(final ExtensionContext context) throws IOException {
+    final var clazz = context.getTestClass().get();
+    if (!isNestedClass(clazz)) {
+      CloseHelper.quietClose(migrator);
+      CloseHelper.quietCloseAll(closables);
+      FileUtil.deleteFolderIfExists(tempDir);
+    }
+  }
+
+  @Override
+  public void beforeAll(final ExtensionContext context) throws Exception {
+    final var clazz = context.getTestClass().get();
+    if (!isNestedClass(clazz)) {
+      indexPrefix = clazz.getSimpleName().toLowerCase();
+      tempDir = Files.createTempDirectory(indexPrefix + "-zeebe");
+      setupDatabase();
+      migrator = new CamundaMigrator(indexPrefix, tempDir, databaseType, databaseUrl);
+      migrator.initialize(initialEnvOverrides);
+      beforeUpgradeConsumer.accept(databaseType, migrator);
+      upgrade(upgradeEnvOverrides);
+    }
+  }
+
+  /**
+   * Allows to provide a consumer that will be called before the upgrade process starts. This can be
+   * used to setup the database with data that is required for the upgrade process.
+   *
+   * @param upgradeConsumer
+   * @return self for chaining
+   */
+  public MigrationITExtension withBeforeUpgradeConsumer(
+      final BiConsumer<DatabaseType, CamundaMigrator> upgradeConsumer) {
+    beforeUpgradeConsumer = upgradeConsumer;
+    return this;
+  }
+
+  public MigrationITExtension withInitialEnvOverrides(final Map<String, String> envOverrides) {
+    initialEnvOverrides = envOverrides;
+    return this;
+  }
+
+  public MigrationITExtension withUpgradeEnvOverrides(final Map<String, String> envOverrides) {
+    upgradeEnvOverrides = envOverrides;
+    return this;
+  }
+
+  private void upgrade(final Map<String, String> envOverrides) {
+
+    awaitImportersFlushed();
+
+    migrator.update(envOverrides);
+    awaitExporterReadiness();
+
+    /* Ingest an 8.8 Record in order to trigger importers empty batch counting */
+    ingestRecordToTriggerImporters(migrator.getCamundaClient());
+
+    awaitImportersFinished();
+  }
+
+  private boolean isNestedClass(final Class<?> currentClass) {
+    return !ModifierSupport.isStatic(currentClass) && currentClass.isMemberClass();
+  }
+
+  /** TODO: Merge this with the {@link CamundaMultiDBExtension} setup */
```

Comment:
👍🏼 Yeah I guess as follow up we can extract some code into helper/utilities to make them more reusable between the extensions